library/kani_macros/src/sysroot.rs

```diff
@@ -0,0 +1,457 @@
+    }
+}
+
+/// Heyristic to determine if this item originated in some kind of `impl`
+fn is_probably_impl_fn(item_fn: &ItemFn) -> bool {
+    let mut vis = SelfDetector(false);
+    vis.visit_signature(&item_fn.sig);
+    vis.0
+}
+
+/// Collect all named identifiers used in the argument patterns of a function.
+struct ArgumentIdentCollector(HashSet<Ident>);
+
+impl ArgumentIdentCollector {
+    fn new() -> Self {
+        Self(HashSet::new())
+    }
+}
+
+impl<'ast> Visit<'ast> for ArgumentIdentCollector {
+    fn visit_pat_ident(&mut self, i: &'ast syn::PatIdent) {
+        self.0.insert(i.ident.clone());
+        syn::visit::visit_pat_ident(self, i)
+    }
+    fn visit_receiver(&mut self, _: &'ast syn::Receiver) {
+        self.0.insert(Ident::new("self", proc_macro2::Span::call_site()));
+    }
+}
+
+/// Applies the contained renaming to everything visited.
+struct Renamer<'a>(&'a HashMap<Ident, Ident>);
+
+impl<'a> VisitMut for Renamer<'a> {
+    fn visit_expr_path_mut(&mut self, i: &mut syn::ExprPath) {
+        if i.path.segments.len() == 1
+            && let Some(p) = i.path.segments.first_mut()
+            && let Some(new) = self.0.get(&p.ident) {
+                p.ident = new.clone();
+        }
+    }
+
+    /// This restores shadowing. Without this we would rename all ident
+    /// occurrences, but not the binding location. This is because our
+    /// [`visit_expr_path_mut`] is scope-unaware.
+    fn visit_pat_ident_mut(&mut self, i: &mut syn::PatIdent) {
+        if let Some(new) = self.0.get(&i.ident) {
+            i.ident = new.clone();
+        }
+    }
+}
+
+/// Does the provided path have the same chain of identifiers as `mtch` (match)
+/// and no arguments anywhere?
+fn matches_path<E>(path: &syn::Path, mtch: &[E]) -> bool
+where
+    Ident: std::cmp::PartialEq<E>,
+{
+    path.segments.len() == mtch.len()
+        && path.segments.iter().all(|s| s.arguments.is_empty())
+        && path.leading_colon.is_none()
+        && path.segments.iter().zip(mtch).all(|(actual, expected)| actual.ident == *expected)
+}
+
+/// The main meat of handling requires/ensures contracts.
+///
+/// Generates a `check_<fn_name>_<fn_hash>` function that assumes preconditions
+/// and asserts postconditions.
+///
+/// Decorates the original function with `#[checked_by =
+/// "check_<fn_name>_<fn_hash>"]
+///
+/// Each clause (requires or ensures) creates its own check function that calls
+/// the prior check function inside. The innermost check function calls a copy
+/// of the originally decorated function. It is a copy, because the compiler
+/// later replaces all invocations of the original function with this check and
+/// that would also apply to the inner check. We need that to be the untouched
+/// function though so we make a copy that will survive the replacement from the
+/// compiler.
+fn requires_ensures_alt(attr: TokenStream, item: TokenStream, is_requires: bool) -> TokenStream {
```

Comment:
After the latest major changes this function got a lot smaller (in terms of code) and I added much more comments so let me know if you still think it should be changed.

Reply:
Just as a matter of organization. I would say it is more readable if you create a composite of functions, instead of having an argument that changes what the function produces. Although there is a bit more redundancy, I believe it's much easier to read something like:
```rust
fn requires(/* args */) {
  let Some(check_fn) = get_check_fn() else { /* original */ return };
  let new_body = /* requires logic */;
  check_fn.finalize(new_body);
}

fn ensures(/* args */) {
  let Some(check_fn) = get_check_fn() } else { /* original */ return };
  let new_body = /* ensures logic */;
  check_fn.finalize(new_body);
}
```